rust/datafusion/src/datasource/memory.rs

```diff
@@ -0,0 +1,163 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! In-memory data source for presenting a Vec<RecordBatch> as a data source that can be
+//! queried by DataFusion. This allows data to be pre-loaded into memory and then repeatedly
+//! queried without incurring additional file I/O overhead.
+
+use std::cell::RefCell;
+use std::rc::Rc;
+use std::sync::Arc;
+
+use arrow::datatypes::Schema;
+use arrow::record_batch::RecordBatch;
+
+use crate::datasource::{RecordBatchIterator, Table};
+use crate::execution::error::Result;
+
+pub struct MemTable {
+    schema: Arc<Schema>,
+    batches: Vec<RecordBatch>,
+}
+
+impl MemTable {
+    pub fn new(schema: Arc<Schema>, batches: Vec<RecordBatch>) -> Self {
```

Comment:
In cpp parlace it's common to check if the batches align with the provided schema. Do you think it'd be useful to do here? https://github.com/apache/arrow/blob/master/cpp/src/arrow/table.cc#L408

I suppose for datasources which are read by Arrow from disk, this won't be an issue, but someone could provide mismatching in-memory data.

Reply:
Good point. For now I just added a check that the column counts match since we don't yet have a way to compare schemas.

Reply:
Also, I filed https://issues.apache.org/jira/browse/ARROW-4727